In [1]:
import pandas as pd
import numpy as np
from scipy import stats 

In [2]:
df=pd.read_csv('https://github.com/HamoyeHQ/HDSC-Introduction-to-Python-for-machine-learning/files/7768140/FoodBalanceSheets_E_Africa_NOFLAG.csv' ,delimiter=',', encoding="latin-1")

In [3]:
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [4]:
df.shape

(60943, 12)

**Q11 What is the total sum of Animal Fat produced in 2014 and 2017 respectively?**

In [5]:
df.groupby(['Item'])[['Y2014','Y2017']].sum('Animal fats')

Y2014      Y2017
Item                                     
Alcohol, Non-Food      2403.00    2348.00
Alcoholic Beverages  102410.11   95581.06
Animal Products       11935.65   11547.65
Animal fats          209460.54  269617.53
Apples and products    9499.23   10198.90
...                        ...        ...
Vegetables, Other    155038.96  157752.59
Vegetal Products     107145.19  107655.20
Wheat and products   232670.13  240047.62
Wine                   4497.36    4178.02
Yams                 200396.96  229174.59

[119 rows x 2 columns]

**Hence the total sum of Animal Fat in Y2014 and Y2017 are 20.9460.54 and 269617.53 respectively**

**Q12 What is the mean and standard deviation across the whole dataset for the year 2015 to 3 decimal places?**

In [6]:
df['Y2015'].agg({'mean','std'}).round(3)

mean     135.236
std     1603.404
Name: Y2015, dtype: float64

**The mean and standard deviation of the year 2015 are '135.236' and '1603.404' respectively**

**Q13 What is the total number and percentage of missing data in 2016 to 2 decimal places?**

In [7]:
missing_data_2016=df['Y2016'].isna().sum()
percent_missing_data_2016=missing_data_2016*100/len(df)

In [8]:
print("The total number of missing data in 2016 is "+ str(missing_data_2016))
print("The total percentage of missing data in 2016 is "+ str(percent_missing_data_2016.round(2)))

The total number of missing data in 2016 is 1535
The total percentage of missing data in 2016 is 2.52


**The Total number of missing data in 2016 is 1535**


**The Total percentage of missing data in 2016 is 2.52**

**Q14 Which year had the highest correlation with ‘Element Code’?**

In [9]:
cor1=df['Element Code'].corr(df['Y2014'])
print("'Y2014' is "+ str(cor1))
cor2=df['Element Code'].corr(df['Y2015'])
print("'Y2015' is "+ str(cor2))
cor3=df['Element Code'].corr(df['Y2016'])
print("'Y2016' is "+ str(cor3))
cor4=df['Element Code'].corr(df['Y2017'])
print("'Y2017' is "+ str(cor4))
cor5=df['Element Code'].corr(df['Y2018'])
print("'Y2018' is "+ str(cor5))

 

'Y2014' is 0.024456819495221733
'Y2015' is 0.02388919930163992
'Y2016' is 0.023443997584961344
'Y2017' is 0.02425400776004579
'Y2018' is 0.024278571953519505


In [10]:
print("The highest correation value is "+ str(max(cor1,cor2,cor3,cor4,cor5).round(4)))
print("Therefore the year with the highest correlation is Y2014 with value of "+ str(cor1.round(4) ))

The highest correation value is 0.0245
Therefore the year with the highest correlation is Y2014 with value of 0.0245


**The Year with the highest correlation is 'Y2014'**

**Q15 What year has the highest sum of Import Quantity?**

In [11]:
df.groupby('Element')[['Y2014','Y2017','Y2016','Y2015','Y2018']].sum('Import Quantity')

Y2014       Y2017       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2088198.10  2044842.70   
Export Quantity                          150020.64   182338.80   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10253.84    10102.77   
Feed                                     216927.89   223705.68   228958.65   
Food                                    1212332.49  1258888.28  1247022.17   
Food supply (kcal/capita/day)            454257.00   454681.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    48690.04    48985.28   
Import Quantity                          274144.48   294559.09   286582.78   
Losses                                   153223.00   160614.00   157787.00   
Other uses (non-food)                     78718.13    91645.97    69563.68   
Processing                               282923.00   292836.00   280631.00   
Production                              1931287.75  2030056.89  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11842.45    11779.69   
Residuals                                 30149.00    35500.00    37224.00   
Seed                                      21922.92    24870.14    23389.20   
Stock Variation                           58749.83    54316.91    33140.12   
Total Population - Both sexes           1031585.00  1112641.00  1085107.00   
Tourist consumption                         416.00       91.00       89.00   

                                             Y2015       Y2018  
Element                                                         
Domestic supply quantity                2021493.55  2161192.10  
Export Quantity                          157614.47   181594.80  
Fat supply quantity (g/capita/day)        10235.74    10258.69  
Feed                                     225050.22   233489.68  
Food                                    1232361.10  1303841.28  
Food supply (kcal/capita/day)            453383.00   455261.00  
Food supply quantity (kg/capita/yr)       49345.13    49056.85  
Import Quantity                          267018.46   287997.09  
Losses                                   155439.00   163902.00  
Other uses (non-food)                     66254.41    91300.97  
Processing                               287929.00   308429.00  
Production                              1947019.39  2075072.89  
Protein supply quantity (g/capita/day)    11833.95    11833.56  
Residuals                                 30045.00    34864.00  
Seed                                      23976.82    25263.14  
Stock Variation                           34910.99    20577.91  
Total Population - Both sexes           1058081.00  1140605.00  
Tourist consumption                         349.00       90.00

**The Year '2017' has the highest import quantity of about 294559.09**

**Q16 What is the total number of the sum of Production in 2014?**

In [12]:
df.groupby(['Element'])[['Y2014']].sum()

Y2014
Element                                           
Domestic supply quantity                1996716.35
Export Quantity                          150020.64
Fat supply quantity (g/capita/day)        10225.56
Feed                                     216927.89
Food                                    1212332.49
Food supply (kcal/capita/day)            454257.00
Food supply quantity (kg/capita/yr)       49650.63
Import Quantity                          274144.48
Losses                                   153223.00
Other uses (non-food)                     78718.13
Processing                               282923.00
Production                              1931287.75
Protein supply quantity (g/capita/day)    11836.46
Residuals                                 30149.00
Seed                                      21922.92
Stock Variation                           58749.83
Total Population - Both sexes           1031585.00
Tourist consumption                         416.00

**The sum of Production in the year 2014 is '1931287.75'**

**Q17 Which of these elements had the highest sum in 2018?**

In [13]:
d=df.groupby(['Y2018'], sort=True)[['Element']].sum().reset_index().sort_values(by=['Y2018'], ascending=[False])
d

Y2018                         Element
4590  195875.0   Total Population - Both sexes
4589  118172.0                      Production
4588  117038.0        Domestic supply quantity
4587  109224.0   Total Population - Both sexes
4586   98424.0   Total Population - Both sexes
...        ...                             ...
4      -1442.0                 Stock Variation
3      -1445.0                 Stock Variation
2      -2156.0  Stock VariationStock Variation
1      -2450.0                 Stock Variation
0      -3396.0                 Stock Variation

[4591 rows x 2 columns]

**The Element with the highest sum is 'Total Population-Both sexes' iwth a value of 195875.0**

**Q18Which of these elements had the 3rd lowest sum in 2018?
Hint-  Select columns ‘Y2018’ and ‘Element’, Perform a groupby operation on ‘Element’ on the selected dataframe and answer the  question.**  

In [14]:
e=df.groupby(['Y2018'], sort=True)['Element'].sum().reset_index().sort_values(by=['Y2018'], ascending=[True])
e

Y2018                         Element
0      -3396.0                 Stock Variation
1      -2450.0                 Stock Variation
2      -2156.0  Stock VariationStock Variation
3      -1445.0                 Stock Variation
4      -1442.0                 Stock Variation
...        ...                             ...
4586   98424.0   Total Population - Both sexes
4587  109224.0   Total Population - Both sexes
4588  117038.0        Domestic supply quantity
4589  118172.0                      Production
4590  195875.0   Total Population - Both sexes

[4591 rows x 2 columns]

**The Element with the 3rd Lowest sum in 2018 is 'Stock VariationStock Variation' which is -2156.0**



**Q19 What is the total Import Quantity in Algeria in 2018?**

In [15]:

import_Algeria=df.groupby(['Y2018','Area'], sort=True)[['Element']].sum().reset_index().sort_values(by=['Area'], ascending=[True])
import_Algeria

Y2018      Area                                            Element
3978      2.14   Algeria                                         Production
3174      1.14   Algeria                                         Production
11588   218.31   Algeria                Food supply quantity (kg/capita/yr)
14178  2322.00   Algeria                                    Import Quantity
4987      4.63   Algeria                Food supply quantity (kg/capita/yr)
...        ...       ...                                                ...
2214      0.51  Zimbabwe             Protein supply quantity (g/capita/day)
10148   100.00  Zimbabwe                          Import QuantityProduction
2189      0.50  Zimbabwe             Protein supply quantity (g/capita/day)
7276     22.00  Zimbabwe  FoodLossesProductionDomestic supply quantityPr...
3935      2.05  Zimbabwe                Food supply quantity (kg/capita/yr)

[14853 rows x 3 columns]

**The Total Import Quantity in Algeria in Year 2018 is '2322.00'**

**Q20 What is the total number of unique countries in the dataset?**

In [16]:
len(df['Area'].unique())

49

**The Total Number of unique countries in the dataset is '49'**

**some of the references include**

**https://stackoverflow.com/questions/43066374/how-to-find-out-number-of-unique-values-in-a-column-along-with-count-of-the-uniq
https://www.datasciencemadesimple.com/mean-variance-and-standard-deviation-of-column-in-pyspark/
https://sparkbyexamples.com/pandas/pandas-groupby-sort-within-groups/#:~:text=2.-,Sort%20within%20Groups%20of%20groupby()%20Result%20in%20DataFrame,of%20rows%20within%20each%20group.
https://stackoverflow.com/questions/44742945/i-applied-sum-on-a-groupby-and-i-want-to-sort-the-values-of-the-last-column
https://stackoverflow.com/questions/51070985/find-out-the-percentage-of-missing-values-in-each-column-in-the-given-dataset**    